File: app/models/lead.rb
# == Schema Information
#
# Table name: leads
#
#  id                  :bigint           not null, primary key
#  courses_data        :text
#  email               :string
#  phone               :string
#  state               :string
#  survey_answers_data :text
#  telegram            :string
#  whatsapp            :string
#  created_at          :datetime         not null
#  updated_at          :datetime         not null
#  ahoy_visit_id       :bigint
#  user_id             :bigint           not null
#  ym_client_id        :string
#
# Indexes
#
#  index_leads_on_ahoy_visit_id  (ahoy_visit_id)
#  index_leads_on_user_id        (user_id)
#
# Foreign Keys
#
#  fk_rails_...  (ahoy_visit_id => ahoy_visits.id)
#  fk_rails_...  (user_id => users.id)
#
class Lead < ApplicationRecord
  belongs_to :user
  belongs_to :ahoy_visit, optional: true, class_name: "Ahoy::Visit"
  serialize :survey_answers_data
  serialize :courses_data

  def self.ransackable_attributes(auth_object = nil)
    [ "created_at", "email", "id", "phone", "state", "telegram", "updated_at", "user_id", "whatsapp" ]
  end
end

File: ent/schema/lead.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/schema/field"
)

// Lead maps the legacy `leads` table (sales leads captured from the site). The
// public lead form writes it (createLead) and the admin list reads it.
//
// Two baseline columns stay unmapped on purpose: `ahoy_visit_id`, because ahoy
// is not ported (ADR-0015) and the first-visit attribution travels in the
// LeadCreated event instead, and `state`, which legacy never set on this path
// and nothing reads. Atlas owns the schema, so omitting them drops nothing.
// The table name already matches ent's default plural of `Lead`, so no @Table
// annotation is needed.
//
// `full_name` is not a column: the legacy serializer derives it from the
// associated user.
type Lead struct {
	ent.Schema
}

// Mixin supplies Rails' application-side timestamps: both columns are NOT NULL
// with no DB default, so ent must fill them on insert like ActiveRecord did.
func (Lead) Mixin() []ent.Mixin {
	return []ent.Mixin{TimestampsMixin{}}
}

func (Lead) Fields() []ent.Field {
	return []ent.Field{
		// NOT NULL in the baseline (FK to users), hence a value field.
		field.Int("user_id"),
		field.String("email").Optional().Nillable(),
		field.String("phone").Optional().Nillable(),
		field.String("whatsapp").Optional().Nillable(),
		field.String("telegram").Optional().Nillable(),
		field.String("survey_answers_data").Optional().Nillable(),
		field.String("courses_data").Optional().Nillable(),
		// The Metrika client id the form carried.
		field.String("ym_client_id").Optional().Nillable(),
	}
}
